src/exception.c

```diff
@@ -281,16 +289,97 @@ void __onCriticalException(volatile reg_block_t* regs)
+ * @brief Register a handler that will be called when the user
+ *        presses the RESET button. 
+ * 
+ * The N64 sends an interrupt when the RESET button is pressed,
+ * and then actually resets the console after about ~500ms (but less
+ * on some models, see #RESET_TIME_LENGTH).
+ * 
+ * Registering a handler can be used to perform a clean reset.
+ * Technically, at the hardware level, it is important that the RCP
+ * is completely idle when the reset happens, or it might freeze
+ * and require a power-cycle to unfreeze. This means that any
+ * I/O, audio, video activity must cease before #RESET_TIME_LENGTH
+ * has elapsed.
+ * 
+ * This entry point can be used by the game code to basically
+ * halts itself and stops issuing commands. Libdragon itself will
+ * register handlers to halt internal modules so to provide a basic
+ * good reset experience.
+ * 
+ * Handlers can use #exception_reset_time to read how much has passed
+ * since the RESET button was pressed.
+ * 
+ * @param cb    Callback to invoke when the reset button is pressed.
+ * 
+ * @note  Reset handlers are called under interrupt.
+ * 
  */
-void __onResetException(volatile reg_block_t* regs)
+void register_reset_handler( void (*cb)(void) )
 {
-	exception_t e;
-	
-	if(!__exception_handler) { return; }
+	for (int i=0;i<MAX_RESET_HANDLERS;i++)
+	{		
+		if (!__prenmi_handlers[i])
+		{
+			__prenmi_handlers[i] = cb;
+			return;
+		}
+	}
+	assertf(0, "Too many pre-NMI handlers\n");
+}
 
-	__fetch_regs(&e, EXCEPTION_TYPE_RESET, regs);
-	__exception_handler(&e);
+/** 
+ * @brief Check whether the RESET button was pressed and how long we are into
+ *        the reset process.
+ * 
+ * This function returns how many ticks have elapsed since the user has pressed
+ * the RESET button, or 0 if the user has not pressed it.
+ * 
+ * It can be used by user code to perform actions during the RESET
+ * process (see #register_reset_handler). It is also possible to simply
+ * poll this value to check at any time if the button has been pressed or not.
+ * 
+ * The reset process takes about 500ms between the user pressing the
+ * RESET button and the CPU being actually reset, though on some consoles
+ * it seems to be much less. See #RESET_TIME_LENGTH for more information.
+ * For the broadest compatibility, please use #RESET_TIME_LENGTH to implement
+ * the reset logic.
+ * 
+ * @return Ticks elapsed since RESET button was pressed, or 0
+ * 
+ * @see register_reset_handler
+ * @see #RESET_TIME_LENGTH
+ */
+uint32_t exception_reset_time( void )
+{
+	if (!__prenmi_tick) return 0;
+	return TICKS_SINCE(__prenmi_tick);
 }
 
+
+/**
+ * @brief Respond to a reset exception.
+ * 
+ * Calls the handlers registered by #register_reset_handler.
+ */
+void __onResetException( volatile reg_block_t* regs )
+{
+	/* This function will be called many times becuase there is no way
+	   to acknowledge the pre-NMI interrupt. So make sure it does nothing
+	   after the first call. */
+	if (__prenmi_tick) return;
```

Comment:
It is theoretically possible to write a 0 ticks so this may execute the callbacks more than one. Not very important or critical but maybe another reason for a negative initial value?

Reply:
This is a bit difference. I can see the value -1 as return value for exception_reset_time, but __prenmi_tick is a uint32_t returned by TICKS_READ so it cannot be negative, it would not work with macros like TICKS_DISTANCE. 

You have a point that there is 1 possibility in 4 billions that libdragon reads the COP0 count when it is exactly 0, and thus misses the reset event. I have fixed this now by always forcing the last bit to 1 when reading. As a side-effect, this will decrease the reset time by 0.5 cycles on average which has no practical effect.